File: Example 7 JNA/src/Example7.java
// Projeto : JNA Example #7
// Classe : Example7.java
// Função : Retrieve an Array of Structs from C
// Autor(es) : Hugo Dionizio Santos
// Data : Dom 10 Mai 2015 20:45:07 BRT 

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.ptr.PointerByReference;		
import com.sun.jna.Pointer;
import com.sun.jna.Structure;
import com.sun.jna.Structure.ByReference;
import com.sun.jna.ptr.IntByReference;

public class Example7 {
	public interface CLibrary extends Library {
		public static class Example7Struct extends Structure {
			public static class ByReference extends Example7Struct implements Structure.ByReference {}
		
			public int val;
		
			public Example7Struct() {}
			public Example7Struct(Pointer p) {
				super(p);
			}
		}

		public void example7_getStructArray(PointerByReference valsRef, IntByReference numValsRef);
		public void example7_cleanupStructArray(Pointer p);
}

	static public void main(String argv[]) {
		CLibrary clib;
		PointerByReference valsRefPtr;

		try {
			clib = (CLibrary)Native.loadLibrary("./libexample7.so", CLibrary.class);
			// 
			valsRefPtr = new PointerByReference();
		}
		catch (UnsatisfiedLinkError e) {
			clib = (CLibrary)Native.loadLibrary("./bin/libexample7.so", CLibrary.class);
			// 
			valsRefPtr = new PointerByReference();
		}

		IntByReference numValsRef = new IntByReference();
		clib.example7_getStructArray(valsRefPtr, numValsRef);
		int numVals = numValsRef.getValue();

		// extract the pointed-to pointer
		Pointer pVals = valsRefPtr.getValue();

		// read the pointed-to memory into an Example7Struct
		CLibrary.Example7Struct valsRef = new CLibrary.Example7Struct(pVals);
		valsRef.read();

		// Structure.toArray copies the data from the C-allocated block of memory
		// into a Java array of Example7Struct objects
		// Note that for large arrays, this can be extremely slow
		CLibrary.Example7Struct[] e7vals = (CLibrary.Example7Struct[])valsRef.toArray(numVals);
		System.out.println("example 7: retrieved " + numVals + " values:");
		for (CLibrary.Example7Struct e7val : e7vals) {
			System.out.println("\t" + e7val.val);
		}

		// call C code to clean up memory allocated in C
		System.out.println("\t(example 7: cleanup)");
		clib.example7_cleanupStructArray(pVals);
	}
}

File: Example 7 JNA/src/example7.c
// Projeto : JNA Example #7
// Algoritmo : example7.c
// Função : Retrieve an Array of Structs from C
// Autor(es) : Hugo Dionizio Santos (adaptações e correções), Ethan Shayne in http://www.eshayne.com/jnaex
// Data : Dom 10 Mai 2015 20:45:07 BRT

#include <stdio.h>
#include <stdlib.h>

// Seção de Variáveis Globais
typedef struct Example7Struct {
	int val;
} Example7Struct;

// Interface do algoritmo

// Função Principal

// Seção de Funções
void example7_getStructArray(Example7Struct** ppVals, int* pNumVals)
{
	*pNumVals = 3;
	*ppVals = (Example7Struct*)malloc(sizeof(Example7Struct) * 3);
	memset(*ppVals, 0, sizeof(Example7Struct) * 3);
	(*ppVals)[0].val = 1;
	(*ppVals)[1].val = 2;
	(*ppVals)[2].val = 3;
}

void example7_cleanupStructArray(Example7Struct* pMem)
{
	printf("(C) cleaning up memory...\n");
	free(pMem);
}

// Fim do algoritmo
